### DISCUSSION TITLE
Missing clip following end message

### DISCUSSION THREAD

**Discussion by ediguidedog** [blakeblackshear/frigate#3525] [Ask A Question] [answered]
Firstly let me say frigate is brilliant. I moved from zoneminder and its a breath of fresh air. However it seems the more I learn about frigate the less I know,  I have a home assistant automation set up to act on the "end" MQTT event message to send the video of any event that involves my front path and this normally works fine. 

Today I returned from a walk with both my wife and the dog and there was no video. Looking in the hass automation logs there were at least 2 person end events sent (I assume one for each person) and they both had differing event IDs but only one of the event IDs actually returned a video. What happened was the both events fired the HA automation but while HA was processing the one with the missing video the one with the actual video was ignored by HA as I have mode set to single 

From the messages I cant actually see any way of understanding which one had the valid event ID that I could use to locate the video.

Questions 
1 is there a way I can find out from the MQTT message whether an event video actually exists apart from calling the API and seeing if a clip is returned?

2 the end event that failed to return a video still shows as having someone in `current_zones`. If that was the case should an end event even be generated in the first place? (or is this the person 1 end event but still showing person 2 in a zone?)

So this was the end event that didn't have a video and the API returns and "Event not found" message 

    before:
      id: 1658747786.403148-ej5glq
      camera: front
      frame_time: 1658747818.648181
      snapshot_time: 1658747818.648181
      label: person
      top_score: 0.84375
      false_positive: false
      start_time: 1658747786.403148
      end_time: null
      score: 0.84375
      box:
        - 1021
        - 157
        - 1182
        - 461
      area: 48944
      ratio: 0.5296052631578947
      region:
        - 668
        - 17
        - 1280
        - 629
      stationary: false
      motionless_count: 0
      position_changes: 2
      current_zones:
        - doorstep
      entered_zones:
        - front path
        - doorstep
      has_clip: true
      has_snapshot: true
    after:
      id: 1658747786.403148-ej5glq
      camera: front
      frame_time: 1658747819.717833
      snapshot_time: 1658747818.648181
      label: person
      top_score: 0.84375
      false_positive: false
      start_time: 1658747786.403148
      end_time: 1658747824.978598
      score: 0.82421875
      box:
        - 1098
        - 222
        - 1244
        - 477
      area: 37230
      ratio: 0.5725490196078431
      region:
        - 684
        - 11
        - 1280
        - 607
      stationary: false
      motionless_count: 1
      position_changes: 2
      current_zones:
        - doorstep
      entered_zones:
        - front path
        - doorstep
      has_clip: true
      has_snapshot: true
    type: end

And this was the one that worked 

    before:
      id: 1658747787.244731-0av36i
      camera: front
      frame_time: 1658747816.966257
      snapshot_time: 1658747805.377687
      label: person
      top_score: 0.84375
      false_positive: false
      start_time: 1658747787.244731
      end_time: null
      score: 0.8046875
      box:
        - 1131
        - 197
        - 1252
        - 502
      area: 36905
      ratio: 0.39672131147540984
      region:
        - 724
        - 45
        - 1280
        - 601
      stationary: false
      motionless_count: 0
      position_changes: 2
      current_zones: []
      entered_zones:
        - front path
        - doorstep
      has_clip: true
      has_snapshot: true
    after:
      id: 1658747787.244731-0av36i
      camera: front
      frame_time: 1658747823.497891
      snapshot_time: 1658747805.377687
      label: person
      top_score: 0.84375
      false_positive: false
      start_time: 1658747787.244731
      end_time: 1658747828.768139
      score: 0.7890625
      box:
        - 1152
        - 195
        - 1257
        - 498
      area: 31815
      ratio: 0.3465346534653465
      region:
        - 808
        - 86
        - 1280
        - 558
      stationary: false
      motionless_count: 19
      position_changes: 2
      current_zones: []
      entered_zones:
        - front path
        - doorstep
      has_clip: true
      has_snapshot: true
    type: end

3 so if the answer is to wait till the end message appears where the `current_zone `is empty, how do I test that `current_zones `in empty? I've tried 
 ` trigger.payload_json["after"]["current_zones"] == "" `
and 
`  trigger.payload_json["after"]["current_zones"][0] == "" `
but neither seem to work 

I guess I could solve the problem by changing the automation mode to allow concurrent running but I'd much rather send one message that is sure to work over a bunch where one will work but the rest will fail

Honestly frigate is brilliant and I've really enjoyed learning about how it works and hope I can contribute on a technical level at some point in the future.

**Comment by NickM-27** [comment]
1. Yes, see [the MQTT docs](https://docs.frigate.video/integrations/mqtt#frigateevents) there is the field `has_clip`
2. If an object goes into a zone and then disappears this can totally happen, the fields from objects aren't traded like that.
3. I wouldn't require that current_zones is empty. You can see [the guide I wrote](https://github.com/blakeblackshear/frigate/discussions/2898) which shows how I do this. It runs in `parallel` mode (if a person and dog go in my yard at similar times I'd want to know about both). Hopefully that can give some idea as to how it works for me.

**Reply by ediguidedog** [reply]
> 



> 1. Yes, see [the MQTT docs](https://docs.frigate.video/integrations/mqtt#frigateevents) there is the field `has_clip`

I thought that but both events above show 'has_clip' as true but only once clip actually exists and it's against the other event ID
so 

https://frigate.mydomain.com/api/events/1658747786.403148-ej5glq/clip.mp4 returns an error 
and 
https://frigate.mydomain.com/api/events/1658747787.244731-0av36i/clip.mp4 returns the clip 

Is this a bug? Should the same event video be reachable via the API from multiple event ID's that happened concurrently or should 'has_clip' have been false for the first event?


> 2. If an object goes into a zone and then disappears this can totally happen, the fields from objects aren't traded like that.

That's probably what happened as I disappeared inside the house 

> 3. I wouldn't require that current_zones is empty. You can see [the guide I wrote](https://github.com/blakeblackshear/frigate/discussions/2898) which shows how I do this. It runs in `parallel` mode (if a person and dog go in my yard at similar times I'd want to know about both). Hopefully that can give some idea as to how it works for me.

Thanks I'll take a look at your code. Much appreciated.

**Reply by NickM-27** [reply]
`has_clip` is set to true based on if recordings are enabled. I am not sure why a recording wouldn't have been created for that event but it's possible based on your zone or other recording filters that it wasn't included. If you look at https://frigate.mydomain.com/api/events/1658747786.403148-ej5glq does it return anything? Perhaps that event was never saved at all.

**Reply by ediguidedog** [reply]
> `has_clip` is set to true based on if recordings are enabled. I am not sure why a recording wouldn't have been created for that event but it's possible based on your zone or other recording filters that it wasn't included. If you look at https://frigate.mydomain.com/api/events/1658747786.403148-ej5glq does it return anything? Perhaps that event was never saved at all.

No. Nothing returned. I think Blake has provided an explanation below. I guess if that's going to be an issue I just need to send both event videos in the hope that one still exists

**Comment by blakeblackshear** [comment, marked as answer]
I see this from time to time too. There is a cleanup job that runs periodically to remove events that begin and end at about the same time as other events. This helps avoid creating 3 completely overlapping events when 3 people walk by together. I suspect that's what you ran into here.
